java.lang.IllegalStateException: unexpected psi: class org.jetbrains.kotlin.psi.KtProperty

this is very similar to #164 and started happening after the 2021-01-11 update.

For the given input:
```
class JavaImpl implements MyInterface {
                public int getX() {
                    return 1;
                }
                public int getY() {
                    return 1;
                }
                public void setY(int value) {
                }
            }
```

```
interface MyInterface {
                val x:Int
                var y:Int
            }
```
calling `findOverridee` on `setY` of `JavaImpl` throws the following exception:

```
Caused by: java.lang.IllegalStateException: unexpected psi: class org.jetbrains.kotlin.psi.KtProperty
	at com.google.devtools.ksp.symbol.impl.UtilsKt.toKSFunctionDeclaration(utils.kt:267)
	at com.google.devtools.ksp.symbol.impl.java.KSFunctionDeclarationJavaImpl.findOverridee(KSFunctionDeclarationJavaImpl.kt:52)
	at androidx.room.compiler.processing.ksp.ResolverExtKt.overrides(ResolverExt.kt:82)
	at androidx.room.compiler.processing.ksp.ResolverExtKt.overrides(ResolverExt.kt:73)
```

--- a/api/src/main/kotlin/com/google/devtools/ksp/symbol/KSFunctionDeclaration.kt
+++ b/api/src/main/kotlin/com/google/devtools/ksp/symbol/KSFunctionDeclaration.kt
@@ -80,10 +80,10 @@ interface KSFunctionDeclaration : KSDeclaration, KSDeclarationContainer {
      * to the current containing declaration is selected. If they are in the same level, the
      * function of the first specified interface (in source) will be returned.
      *
-     * @return [KSFunctionDeclaration] for the original function, if overriding, otherwise null.
+     * @return [KSDeclaration] for the original declaration, if overriding, otherwise null.
      * Calling [findOverridee] is expensive and should be avoided if possible.
      */
-    fun findOverridee(): KSFunctionDeclaration?
+    fun findOverridee(): KSDeclaration?
 
     /**
      * Returns the type of the [function] when it is viewed as member of the [containing] type.

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/binary/KSFunctionDeclarationDescriptorImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/binary/KSFunctionDeclarationDescriptorImpl.kt
@@ -28,7 +28,7 @@ import com.google.devtools.ksp.symbol.*
 import com.google.devtools.ksp.symbol.impl.KSObjectCache
 import com.google.devtools.ksp.symbol.impl.findClosestOverridee
 import com.google.devtools.ksp.symbol.impl.toFunctionKSModifiers
-import com.google.devtools.ksp.symbol.impl.toKSFunctionDeclaration
+import com.google.devtools.ksp.symbol.impl.toKSDeclaration
 import com.google.devtools.ksp.symbol.impl.toKSModifiers
 import org.jetbrains.kotlin.descriptors.ClassConstructorDescriptor
 import org.jetbrains.kotlin.load.java.isFromJava
@@ -42,9 +42,8 @@ class KSFunctionDeclarationDescriptorImpl private constructor(val descriptor: Fu
         fun getCached(descriptor: FunctionDescriptor) = cache.getOrPut(descriptor) { KSFunctionDeclarationDescriptorImpl(descriptor) }
     }
 
-    override fun findOverridee(): KSFunctionDeclaration? {
-        val descriptor = ResolverImpl.instance.resolveFunctionDeclaration(this)
-        return descriptor?.findClosestOverridee()?.toKSFunctionDeclaration()
+    override fun findOverridee(): KSDeclaration? {
+        return descriptor?.findClosestOverridee()?.toKSDeclaration()
     }
 
     override val typeParameters: List<KSTypeParameter> by lazy {

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSClassDeclarationJavaImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSClassDeclarationJavaImpl.kt
@@ -34,7 +34,6 @@ import com.google.devtools.ksp.symbol.impl.kotlin.KSNameImpl
 import com.google.devtools.ksp.symbol.impl.kotlin.getKSTypeCached
 import com.google.devtools.ksp.symbol.impl.replaceTypeArguments
 import com.google.devtools.ksp.symbol.impl.synthetic.KSConstructorSyntheticImpl
-import com.google.devtools.ksp.symbol.impl.toKSFunctionDeclaration
 import com.intellij.psi.PsiEnumConstant
 import org.jetbrains.kotlin.descriptors.DescriptorVisibilities
 import org.jetbrains.kotlin.descriptors.PropertyDescriptor

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSFunctionDeclarationJavaImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/java/KSFunctionDeclarationJavaImpl.kt
@@ -26,7 +26,6 @@ import com.google.devtools.ksp.symbol.impl.kotlin.KSNameImpl
 import com.intellij.lang.jvm.JvmModifier
 import com.intellij.psi.PsiJavaFile
 import com.intellij.psi.PsiMethod
-import com.intellij.psi.impl.source.PsiClassReferenceType
 
 class KSFunctionDeclarationJavaImpl private constructor(val psi: PsiMethod) : KSFunctionDeclaration, KSDeclarationJavaImpl(psi),
     KSExpectActual by KSExpectActualNoImpl() {
@@ -48,9 +47,9 @@ class KSFunctionDeclarationJavaImpl private constructor(val psi: PsiMethod) : KS
         KSFileJavaImpl.getCached(psi.containingFile as PsiJavaFile)
     }
 
-    override fun findOverridee(): KSFunctionDeclaration? {
+    override fun findOverridee(): KSDeclaration? {
         val descriptor = ResolverImpl.instance.resolveFunctionDeclaration(this)
-        return descriptor?.findClosestOverridee()?.toKSFunctionDeclaration()
+        return descriptor?.findClosestOverridee()?.toKSDeclaration()
     }
 
     override val declarations: Sequence<KSDeclaration> = emptySequence()

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/kotlin/KSFunctionDeclarationImpl.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/kotlin/KSFunctionDeclarationImpl.kt
@@ -38,9 +38,9 @@ class KSFunctionDeclarationImpl private constructor(val ktFunction: KtFunction) 
         fun getCached(ktFunction: KtFunction) = cache.getOrPut(ktFunction) { KSFunctionDeclarationImpl(ktFunction) }
     }
 
-    override fun findOverridee(): KSFunctionDeclaration? {
+    override fun findOverridee(): KSDeclaration? {
         val descriptor = ResolverImpl.instance.resolveFunctionDeclaration(this)
-        return descriptor?.findClosestOverridee()?.toKSFunctionDeclaration()
+        return descriptor?.findClosestOverridee()?.toKSDeclaration()
     }
 
     override val simpleName: KSName by lazy {

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/symbol/impl/utils.kt
@@ -314,7 +314,7 @@ internal fun KotlinType.replaceTypeArguments(newArguments: List<KSTypeArgument>)
         TypeProjectionImpl(variance, type)
     })
 
-internal fun FunctionDescriptor.toKSFunctionDeclaration(): KSFunctionDeclaration {
+internal fun FunctionDescriptor.toKSDeclaration(): KSDeclaration {
     if (this.kind != CallableMemberDescriptor.Kind.DECLARATION) return KSFunctionDeclarationDescriptorImpl.getCached(this)
     val psi = this.findPsi() ?: return KSFunctionDeclarationDescriptorImpl.getCached(this)
     // Java default constructor has a kind DECLARATION of while still being synthetic.
@@ -324,6 +324,7 @@ internal fun FunctionDescriptor.toKSFunctionDeclaration(): KSFunctionDeclaration
     return when (psi) {
         is KtFunction -> KSFunctionDeclarationImpl.getCached(psi)
         is PsiMethod -> KSFunctionDeclarationJavaImpl.getCached(psi)
+        is KtProperty -> KSPropertyDeclarationImpl.getCached(psi)
         else -> throw IllegalStateException("unexpected psi: ${psi.javaClass}")
     }
 }

--- a/compiler-plugin/src/test/kotlin/com/google/devtools/ksp/processor/OverrideeProcessor.kt
+++ b/compiler-plugin/src/test/kotlin/com/google/devtools/ksp/processor/OverrideeProcessor.kt
@@ -22,6 +22,7 @@ import com.google.devtools.ksp.getClassDeclarationByName
 import com.google.devtools.ksp.processing.Resolver
 import com.google.devtools.ksp.symbol.KSAnnotated
 import com.google.devtools.ksp.symbol.KSClassDeclaration
+import com.google.devtools.ksp.symbol.KSDeclaration
 import com.google.devtools.ksp.symbol.KSFunctionDeclaration
 import com.google.devtools.ksp.symbol.KSPropertyDeclaration
 
@@ -42,6 +43,7 @@ class OverrideeProcessor: AbstractTestProcessor() {
         logSubject(resolver, "ConflictingSubject4")
         logSubject(resolver, "OverrideOrder1")
         logSubject(resolver, "OverrideOrder2")
+        logSubject(resolver, "JavaAccessorImpl")
         return emptyList()
     }
 
@@ -57,17 +59,27 @@ class OverrideeProcessor: AbstractTestProcessor() {
     private fun logClass(subject: KSClassDeclaration) {
         subject.declarations.filterIsInstance<KSPropertyDeclaration>()
             .forEach {
-                val signature = it.toSignature()
-                val overrideeSignature = it.findOverridee()?.toSignature()
-                results.add("$signature -> $overrideeSignature")
+                checkOverridee(it)
             }
         subject.declarations.filterIsInstance<KSFunctionDeclaration>()
             .filterNot { it.simpleName.asString() in IGNORED_METHOD_NAMES }
             .forEach {
-                val signature = it.toSignature()
-                val overrideeSignature = it.findOverridee()?.toSignature()
-                results.add("$signature -> $overrideeSignature")
+                checkOverridee(it)
             }
+    }
+
+    private fun checkOverridee(declaration: KSDeclaration) {
+        val signature = if (declaration is KSPropertyDeclaration) declaration.toSignature() else (declaration as KSFunctionDeclaration).toSignature()
+        val overrideeSignature = if (declaration is KSPropertyDeclaration) declaration.findOverridee()?.toSignature() else (declaration as KSFunctionDeclaration).findOverridee()?.toSignature()
+        results.add("$signature -> $overrideeSignature")
+    }
+
+    private fun KSDeclaration.toSignature(): String {
+        return when(this) {
+            is KSFunctionDeclaration -> this.toSignature()
+            is KSPropertyDeclaration -> this.toSignature()
+            else -> throw IllegalStateException()
+        }
     }
 
     private fun KSFunctionDeclaration.toSignature(): String {
@@ -99,20 +111,3 @@ class OverrideeProcessor: AbstractTestProcessor() {
     }
 }
 
-interface MyInterface {
-    fun openFoo(): Int { return 1}
-    fun absFoo(): Unit
-}
-
-interface MyInterface2 {
-    fun absFoo(): Unit
-}
-
-abstract class MyAbstract: MyInterface {
-    override fun absFoo(): Unit {val a = 1}
-    override fun openFoo(): Int { return 2 }
-}
-
-class Subject2: MyInterface, MyAbstract() {
-    override fun absFoo(): Unit = TODO()
-}

--- a/compiler-plugin/testData/api/overridee.kt
+++ b/compiler-plugin/testData/api/overridee.kt
@@ -81,6 +81,10 @@
 // OverrideOrder1.foo() -> GrandBaseInterface2.foo()
 // OverrideOrder2:
 // OverrideOrder2.foo() -> GrandBaseInterface1.foo()
+// JavaAccessorImpl:
+// JavaAccessorImpl.getX() -> KtInterfaceWithProperty.x
+// JavaAccessorImpl.getY() -> KtInterfaceWithProperty.y
+// JavaAccessorImpl.setY(value:Int) -> KtInterfaceWithProperty.y
 // END
 // MODULE: lib
 // FILE: lib.kt
@@ -136,6 +140,11 @@ class NoOverride(val propInParam: Int) {
     fun func(val param: Int) {
 
     }
+}
+
+interface KtInterfaceWithProperty {
+    val x:Int
+    var y:Int
 }
 
 abstract class GrandBase {
@@ -272,5 +281,17 @@ public class JavaSubject {
         static String staticMethod() {
             return null;
         }
+    }
+}
+
+// FILE: JavaImpl.java
+class JavaAccessorImpl implements KtInterfaceWithProperty {
+    public int getX() {
+        return 1;
+    }
+    public int getY() {
+        return 1;
+    }
+    public void setY(int value) {
     }
 }

